include/oneapi/dpl/pstl/hetero/dpcpp/unseq_backend_sycl.h

```diff
@@ -202,103 +202,138 @@ struct __init_processing
+                    scalar_reduction_remainder(__base_idx, __adjusted_n, __vec_size_minus_one, __res, __acc...);
+                else
+                    break;
+            }
+        }
+        // Scalar remainder
         else if (__adjusted_global_id < __adjusted_n)
         {
-            const _Size __items_to_process =
-                std::max(((__adjusted_n - __adjusted_global_id - 1) / __stride) + 1, static_cast<_Size>(0));
             new (&__res.__v) _Tp(__unary_op(__adjusted_global_id, __acc...));
-            for (_Size __i = 1; __i < __items_to_process; ++__i)
-                __res.__v = __binary_op(__res.__v, __unary_op(__adjusted_global_id + __stride * __i, __acc...));
+            const _Size __adjusted_global_id_plus_one = __adjusted_global_id + 1;
+            constexpr _Size __vec_size_minus_two = _VecSize - 2;
+            scalar_reduction_remainder(__adjusted_global_id_plus_one, __adjusted_n, __vec_size_minus_two, __res,
+                                       __acc...);
         }
-        return;
-    }
-
-    // For non-SPIR-V targets, we check if the operator is commutative before selecting the appropriate codepath.
-    // On SPIR-V targets, the sequential implementation with the non-commutative operator is currently more performant.
-    static constexpr bool __use_nonseq_impl = !oneapi::dpl::__internal::__is_spirv_target_v && _Commutative::value;
-
-    template <typename _NDItemId, typename _Size, typename _Res, typename... _Acc>
-    inline void
-    operator()(const _NDItemId& __item_id, const _Size& __n, const _Size& __global_offset, _Res& __res,
-               const _Acc&... __acc) const
-    {
-        if constexpr (__use_nonseq_impl)
-            return nonseq_impl(__item_id, __n, __global_offset, __res, __acc...);
-        else
-            return seq_impl(__item_id, __n, __global_offset, __res, __acc...);
     }
 
     template <typename _Size>
     _Size
-    output_size(const _Size& __n, const ::std::uint16_t& __work_group_size) const
+    output_size(const _Size __n, const _Size __work_group_size, const _Size __iters_per_work_item) const
     {
-        if constexpr (__use_nonseq_impl)
+        if (__iters_per_work_item == 1)
+            return __n;
+        if constexpr (_Commutative{})
         {
             _Size __items_per_work_group = __work_group_size * __iters_per_work_item;
             _Size __full_group_contrib = (__n / __items_per_work_group) * __work_group_size;
             _Size __last_wg_remainder = __n % __items_per_work_group;
-
-            _Size __last_wg_contrib = ::std::min(__last_wg_remainder, static_cast<_Size>(__work_group_size));
+            // Adjust remainder and wg size for vector size
+            _Size __last_wg_vec = oneapi::dpl::__internal::__dpl_ceiling_div(__last_wg_remainder, _VecSize);
+            _Size __wg_vec_size = __work_group_size * _VecSize;
+            _Size __last_wg_contrib = std::min(__last_wg_vec, __wg_vec_size);
```

Comment:
`__last_wg_contrib` seems incorrect, we are dividing the remainder by the vector size, and then comparing that vs a full number of elements processed by a full group.  It seems like one half of the min operation is in "vector" space, and the other half is in "element" space.  

I haven't fully wrapped my head around whether we should be multiplying again by `_VecSize` and this is a rounding up to the nearest multiple of the vector size, or if we shouldn't be dividing in the first place. I think perhaps we shouldn't be dividing the remainder by the vector size?

I'm probably missing something.

Reply:
Good catch. I believe `__work_group_size` is the right minimum value to check here.

The general idea is to find how many work-items reduce at least a single element in the coalesced access. Each participating work-item does vector width operation. If each work-item in the last group participates, we have work-grup size partial results. If less than that, we need to divide and round up by vector size to get the busy work-items.

Reply:
That makes sense, and I think it is the appropriate fix.